README.md
Answers Cloud Services API Client Library - PHP
===================
Helper library for connecting to the Answers Cloud Services (ForeSee in particular) web API in a headless manner. You can use this to simplify connecting to the ACS api without requiring a browser or user interaction to grant access to a particular account.
###Version Support
Supports PHP v5.2 and above 

###Installation 
+ Download client library
+ Install extracted library to a sub-path of your codebase

###Simple Usage
```php
require_once realpath('path/to/library/AcsClient.php');

$opts = array(
  consumerKey: "[CONSUMER KEY PROVIDED BY FORESEE]",
  consumerSecret: "[CONSUMER SECRET PROVIDED BY FORESEE]",
  consumerType: "[CONSUMER TYPE PROVIDED BY FORESEE]",
  username: "[USERNAME]",
  password: "[PASSWORD]"
);

/**
 * Set up the client
 */
$client = new AcsClient($opts);

// Call resource endpoint
$client->callResource(
    "currentUser", // Resource endpoint 
    "GET", // HTTP method
    [], // Extra Parameters necessary for some end points 
    function($result) { // callback function to be executed on error and success
        if ($result->isError()) {
            // Preform actions when errors occur
        } else {
            // Preform actions for a successful execution
            $last = $result->getLastResponse();
        }
    }
);
```
Note: Native anonymous function implementation above requires PHP5.3, see http://php.net/manual/en/functions.anonymous.php

###Errors
The ACS Client class throws InvalidArgumentException, see http://php.net/manual/en/class.invalidargumentexception.php, when creating instances or executing methods with missing or invalid argumnets.

Otherwise, ACS Client returns a FALSE when errors occur after setting internal error variables. If not monitoring the return result, errors can be checked by executing isError() and retrieving the errors by calling getErrors().  

If there is a problem authenticating or reaching an endpoint, or if a request is malformed, an error will be generated and stored internally within the ACS Client instance. Possible error codes:
* `INVALIDREQUESTTOKEN` - Could not get a request token. There may be something wrong with your consumer key or consumer secret.
* `COULDNOTLOGIN` - There was a problem with the login process. Probably not due to invalid credentials.
* `INVALIDCREDENTIALS` - Could not log in with the provided credentials.
* `COULDNOTAUTHTOKEN` - Could not authorize the auth token.
* `COULDNOTFINDVERIFIER` - There was a problem with the authentication flow. Might be due to an invalid `consumer_type`, `consumer_key` or `consumer_secret`.
* `COULDNOTGETACCESSTOKEN` - There was a problem with the authentication flow. Might be due to an invalid `consumer_type`, `consumer_key` or `consumer_secret`.
* `COULDNOTGETACCESSTOKENNULL` - There was a problem with the authentication flow. Might be due to an invalid `consumer_type`, `consumer_key` or `consumer_secret`.
* `INVALIDREQUEST` - The request failed for an unknown or generic reason. Please try again.
* `403` - You do not have access to that endpoint with that criteria.
* `404` - The endpoint requested could not be found.

Errors are provided as a simple array with keys 'code' and 'msg'. Example:
```php
[
  "code": "COULDNOTGETACCESSTOKENNULL",
  "msg": "Error getting the access token since they were null."
]
```
###Authentication
Authentication will occur automatically, when necessary, as you make service calls. Therefor, if you have not authenticated already and not yet been granted an access token nor an access token secret, the first call to an endpoint will take longer than the rest since it has to go through the authentication flow first.

If you want to check your credentials or just ensure you're actually able to contact an endpoint using the information provided, you can call the `authenticate()` method:
 
```php
require_once realpath('path/to/library/AcsClient.php');

$opts = array(
  consumerKey: "[CONSUMER KEY PROVIDED BY FORESEE]",
  consumerSecret: "[CONSUMER SECRET PROVIDED BY FORESEE]",
  consumerType: "[CONSUMER TYPE PROVIDED BY FORESEE]",
  username: "[USERNAME]",
  password: "[PASSWORD]"
);

/**
 * Set up the client
 */
$client = new AcsClient($opts);
if ($client->authenticate()) {
    // Authenticated
} else {
    // Error state
}
```

###Accessing without Credentials
ACS uses an oAuth authentication scheme so you may want to take advantage of the fact that once you have an access token and an access token secret, you should not need to keep user credentials around to continue using the service layer. ACS can issue you a long-lived token if you request it so you can just set the following attributes on the `options` object rather than the u/p combo:
```php
// Authenticating without usernames or passwords
require_once realpath('path/to/library/AcsClient.php');

$opts = array(
  consumerKey: "[CONSUMER KEY PROVIDED BY FORESEE]",
  consumerSecret: "[CONSUMER SECRET PROVIDED BY FORESEE]",
  consumerType: "[CONSUMER TYPE PROVIDED BY FORESEE]",
  accessToken: "[AUTHENTICATED ACCESS TOKEN]",
  accessTokenSecret: "[AUTHENTICATED ACCESS TOKEN SECRET]"
);

/**
 * Set up the client
 */
$client = new AcsClient($opts);

$client->callResource(
    "currentUser", // Resource endpoint 
    "GET", // HTTP method
    [], // Extra Parameters necessary for some end points 
    function($result) { // callback function to be executed on error and success
        if ($result->isError()) {
            // Preform actions when errors occur
        } else {
            // Preform actions for a successful execution
            $last = $result->getLastResponse();
        }
    }
);
```
Using this technique can result in performance improvements since it eliminates the need to move through the authentication flow. The only downside is, if your token has expired, then you will need to authenticate again with your username and password.

Getting your access token and secret is easy: just look for them on the `client.opts` object:
```php
/**
 * @var array $accessToken ['accessToken' => '..', 'accessTokenSecret' => '..']
 */
$accessToken = $client->getAccessToken();
print_r($accessToken);
```
###Date Criteria
Many of the endpoints accept dates or date ranges as filtering criteria. The full documentaton for how to format these can be found at the ACS developer portal (http://bit.ly/15uYi0k). We've provided a convenience method to help generate these terse date objects:
```php
$clientID = "[ASSIGNED CLIENTID]";

$client->callResource(
    "endPointUri", // Resource endpoint 
    "GET", // HTTP method
    [
        "criteria" => [
            "dateRange" => $client->getDateObject($clientId, "WEEKTODATE")
        ]
    ],  
    function($result) { // callback function to be executed on error and success
        if ($result->isError()) {
            // Preform actions when errors occur
        } else {
            // Preform actions for a successful execution
            $last = $result->getLastResponse();
        }
    }
);
```
The first argument to `getDateObject()` is always the client id (Number). This is used to handle some custom calendaring functions like fiscal calendar. The rest of the arguments depend on what you want. The second argument can be one of the following constants, for example:
* `WEEKTODATE` - The current week of data.
* `LASTWEEK` - Last week's data.
* `YESTERDAY` - Yesterday's data. Note there is no "today" currently.
* `MONTHTODATE` - The current month of data.
* `LASTMONTH` - Last month
* `QUARTERTODATE` - The current fiscal quarter of data
* `LASTQUARTER` - The last fiscal quarter of data
* `YEARTODATE` - The current year of data.
* `LASTYEAR` - Last year

You can also construct more complex ranges that involve several arguments. Here are come examples to get you started:
```php
$client->getDateObject($clientId, "LAST", 3, "DAYS");
$client->getDateObject($clientId, "LAST", 6, "MONTHS");
$client->getDateObject($clientId, "LAST", 2, "YEARS");
```
Of course you can also get date objects for specific dates and date ranges:
```php
$client->getDateObject($clientId, "2013-03-04");  // Get for April 4, 2013 by string
$client->getDateObject($clientId, new \DateTime("2013-03-04");  // Get for April 4, 2013 by object

// Get for a range between April 4, 2013 and January 1, 2014:
$client->getDateObject($clientId, "2014-03-04", "2015-01-01");  
```
###Fiscal Calendars
Normally, relative and general dates are calculated using normal calendar (Gregorian) dates. If your account is set up with fiscal calendars, you can switch to this instead by passing `FISCAL` as the last argument:
```php
$client->getDateObject($clientId, "LAST", 2, "YEARS", "FISCAL");
```
###Date Comparisons
To perform a date comparison on a period period, which is possible on some endpoints, make the last argument in your `getDateObject()` call equal to `PRIORPERIOD`. Then, assign this to the `dateRangeCompare` attribute of the criteria object:
```php
$client->callResource(
    "endPointUri", // Resource endpoint 
    "GET", // HTTP method
    [
        "criteria" => [
            "dateRange" => $client->getDateObject($clientId, "WEEKTODATE"),
            "dateRangeCompare" => $client->getDateObject($clientId, "WEEKTODATE", "PRIORPERIOD")
        ]
    ],  
    function($result) { // callback function to be executed on error and success
        if ($result->isError()) {
            // Preform actions when errors occur
        } else {
            // Preform actions for a successful execution
            $last = $result->getLastResponse();
        }
    }
);
```

CHANGELOG
# Change Log
All notable changes to this project will be documented in this file.

## [0.0.1+1] - 2015-11-09
### Updated
- Completed unit tests for AcsClient
### Added
- Unit tests for AcsDate

## [0.0.1] - 2015-11-09
### Updated
- Reorganized source code to account for unit tests
### Added
- Unit tests for AcsClient library

## [0.0.0+3] - 2015-11-07
### Updated
- Additional PHPDOC documentation for classes and licenses
- Removed deprecated functions
- Renamed class ForeseeDate to AcsDate

## [0.0.0+2] - 2015-10-30
### Added
- Additional error codes for error states

## [0.0.0+1] - 2015-10-26
### Added
- README.md file describing usage
- Ability to force authentication process with `authenticate()` call
- Ability to set access key and associated secrets to skip the authentication process
- Ability to retrieve access keys and associated secrets for storage and reuse
### Updated
- Function PHPDOC

## [0.0.0] - 2015-10-15
### Added
- Changelog file
- Initial ACS Library Entry Point
- OAuth Library to handle back ground interactions


tests/bootstrap.php
<?php
/**
 * Bootstrap the tests
 *
 * PHP version 5.2
 *
 * @category   OAuthTest
 * @author     Pieter Hordijk <info@pieterhordijk.com>
 * @author     David Desberg  <david@daviddesberg.com>
 * @copyright  Copyright (c) 2012 Pieter Hordijk
 * @license    http://www.opensource.org/licenses/mit-license.html  MIT License
 */
namespace AcsTest;

/**
 * Setting up the default timezone. because well... PHP sucks
 */
date_default_timezone_set('Etc/UTC');

/**
 * Simple SPL autoloader for the OAuthTest mocks
 *
 * @param string $class The class name to load
 *
 * @return void
 */
spl_autoload_register(function ($class) {
    $nslen = strlen(__NAMESPACE__);
    if (substr($class, 0, $nslen) !== __NAMESPACE__) {
        return;
    }
    $path = substr(str_replace('\\', '/', $class), $nslen);
    $path = __DIR__ . $path . '.php';
    if (file_exists($path)) {
        require $path;
    }
});

src/AcsClient.php
<?php
/**
 * The MIT License (MIT)
 * Copyright (c) 2015 Answers Cloud Services
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 * @copyright Answers Cloud Services
 * @author Matthew Stuart <matthew.stuart@answers.com>
 */

/**
 * The ACS API Client Library
 * Provides a library for interaction abstraction with the ACS API {@see http://bit.ly/15uYi0k}
 */
class AcsClient
{
    /**
     * Signature Method - HMAC-SHA1
     * @var string
     */
    const SIGNATURE_METHOD_HMACSHA1 = "HMAC-SHA1";

    /**
     * Default construction passes the optional $config variable to the option parser
     * @param array $config OPTIONAL array of options
     */
    public function __construct(array $config = array())
    {        
        $this->setOptions($config);
    }

    /**
     * Sets internal options from provided array. Will use individual setters if option key matches method name
     *
     * @param array $options
     * @return $this
     */
    public function setOptions(array $options)
    {
        foreach ($options as $key => $value) {
            $method = "set" . ucfirst($key);
            if (method_exists($this, $method)) {
                $this->{$method}($value);
            } else {
                $this->_options[$key] = $value;
            }
        }
        return $this;
    }

    /**
     * Magic function implements proxy "get*" functions
     *
     * @param string $name
     * @param array $arguments
     * @return mixed
     * @throws InvalidArgumentException
     */
    public function __call($name, $arguments)
    {
        if (substr($name, 0, 3) === "get") {
            $attribute = "_" . ($property = lcfirst(substr($name, 3)));
            if (property_exists($this, $attribute)) {
                return $this->{$attribute};
            }
            if (array_key_exists($property, $this->_options)) {
                return $this->_options[$property];
            }
        }
        throw new InvalidArgumentException("Method not defined '$name'");
    }

    /**
     * Primary entry point
     * Function handles the interaction with the ACS API including necessary authentication steps should they be required.
     * This function accepts multiple parameter patterns:
     * <pre>
     *      callResource(string $resource)
     *      callResource(string $resource, function $callback)
     *      callResource(string $resource, string $method)
     *      callResource(string $resource, string $method, array $data)
     *      callResource(string $resource, string $method, function $callback)
     *      callResource(string $resource, array $data, function $callback)
     *      callResource(string $resource, string $method, array $data, function $callback)
     * </pre>
     *
     * @param string $resource URI for ACS API endpoint
     * @param string $method HTTP Method to use for communication, Default: GET
     * @param array $data Optional extra data to provide during the API request
     * @param callable $callback Callback to be executed on completion o the request. Called on success and failure
     * @throws InvalidArgumentException
     * @return $this|bool When the request fails, FALSE is returned otherwise an instance of $this for chaining
     */
    public function callResource($resource, $method = "", $data = null, $callback = null)
    {
        if (!is_string($resource) || !$resource) {
            throw new \InvalidArgumentException("Unexpected \$resource provided");
        }

        $defaultMethod = "GET";
        if (is_callable($method)) {
            $callback = $method;
            $data = array();
            $method = $defaultMethod;
        }
        if (is_callable($data)) {
            $callback = $data;
            $data = array();
        }
        if (is_array($method)) {
            $data = $method;
            $method = $defaultMethod;
        }

        if (
            !is_string($method) ||
            !($method = strtoupper(trim($method)) ?: $defaultMethod) ||
            !in_array($method, array("GET", "POST", "DELETE", "PUT"))
        ) {
            throw new InvalidArgumentException("Invalid supplied method");
        }

        $data = $data ?: array();
        if (!is_array($data)) {
            throw new InvalidArgumentException("Expected array for extra data");
        }

        if (!$callback || !is_callable($callback)) {
            $callback = function() { /* no op */ };
        }

        /* Clean the path */
        $resource = str_replace("\\", "/", $resource);
        if ($resource[0] == "/") {
            $resource = substr($resource, 1);
        }

        if (!$this->_establishAuthorization()) {
            $callback($this);
            return false;
        }

        $oa = $this->_getOAuth();
        $response = $oa->request(
            $resource,
            $method,
            null,
            array()
        );

        $info = $oa->getLastRequest();
        if (!(is_array($info) && isset($info['info']) && isset($info['info']['http_code']) && $info['info']['http_code'] == 200)) {
            // Error State
            $info = (array)$info;
            $http_code = 0;
            if (isset($info['info']) && isset($info['info']['http_code'])) {
                $http_code = $info['info']['http_code'];
            }
            switch ($http_code) {
                case 404:
                    $this->setError("404", "Unknown resource '$resource'");
                    break;
                case 403:
                    $this->setError("403", "Access denied to '$resource'");
                    break;
                default:
                    $this->setError("INVALIDREQUEST", "Request to '$resource' could not complete. Please try again.");
                    break;
            }
        } else if (!is_string($response) || !$response = json_decode($response, true)) {
            $this->setError("INVALIDREQUEST", "Request to '$resource' could not complete. Please try again.");
        }

        if ($this->isError()) {
            $callback($this);
            return false;
        }

        $this->setError(false);
        $this->setLastResponse($response);
        $callback($this);

        return $this;
    }

    /**
     * Helper function to retrieve a properly formatted DATE reference object. This date object is formatted in a way
     * that the ACS API can understand the request properly, it supports moment in time and time range options.
     *
     *
     * @param string $clientId Customer client ID used for fiscal based dates
     * @param mixed $date,... Date operators and operands see readme
     * @return array|bool FALSE on error, array of date details when successful
     */
    public function getDateObject($clientId, $date)
    {
        $arguments = func_get_args();

        // Remove the client id
        array_shift($arguments);
        if (!count($arguments) || !$from = array_shift($arguments)) {
            throw new \InvalidArgumentException("Invalid arguments, date creation requires at least one date parameter");
        }

        $to = $style = $fiscal = $option = null;
        if ($arguments) {
            if (is_object($fiscal = end($arguments))) {
                $fiscal = null;
            }
            if (count($arguments) > 4) {
                $arguments = array_slice($arguments, 0, 4);
            }
            switch (count($arguments)) {
                case 4:
                    list($to, $style, $option, $fiscal) = $arguments;
                    break;
                case 3:
                    list($to, $style, $option) = $arguments;
                    break;

                case 2:
                    list($to, $style) = $arguments;
                    $option = $style;
                    break;
                case 1:
                    $to = $arguments[0];
                    $option = $style = $to;
                    break;
            }
        }

        if (!class_exists('AcsDate', false)) {
            require_once __DIR__ . '/AcsDate.php';
        }
        return \AcsDate::create($clientId, $from, $to, $style, $fiscal, $option);
    }

    /**
     * Setter: Sets internal username for auth
     * @param string $username
     * @return $this
     */
    public function setUsername($username)
    {
        $this->_username = $username;
        return $this;
    }

    /**
     * Setter: Sets internal password for auth
     * @param $password
     * @return $this
     */
    public function setPassword($password)
    {
        $this->_password = $password;
        return $this;
    }

    /**
     * Setter: Sets internal consumer key parameter
     * @param $key
     * @return $this
     */
    public function setConsumerKey($key)
    {
        $this->_consumerKey = $key;
        return $this;
    }

    /**
     * Setter: Sets internal consumer secret
     * @param $secret
     * @return $this
     */
    public function setConsumerSecret($secret)
    {
        $this->_consumerSecret = $secret;
        return $this;
    }

    /**
     * Seter: Sets internal last response instance
     * @param $response
     * @return $this;
     */
    public function setLastResponse($response)
    {
        $this->_lastResponse = $response;
        return $this;
    }

    /**
     * Getter: Retrieves last response instance
     * @return null
     */
    public function getLastResponse()
    {
        return $this->_lastResponse;
    }

    /**
     * Setter: Sets internal error tracker
     *
     * @param string $code Short code, see readme
     * @param string $message Long message description indicating error
     * @return $this
     */
    public function setError($code, $message = "")
    {
        if ($code) {
            $code = array(
                'code' => $code,
                'message' => $message
            );
        }
        $this->_error = $code;
        return $this;
    }

    /**
     * State: Indicates if current error state
     * @return bool
     */
    public function isError()
    {
        return !!$this->_error;
    }

    /**
     * Getter: Retrieves the current error instance
     * @return bool
     */
    public function getError()
    {
        return $this->_error;
    }

    /**
     * Getter: retrieves the currently set signature method or {@see SIGNATURE_METHOD_HMACSHA1} by default
     * @return string
     */
    public function getSignatureMethod()
    {
        return $this->_signatureMethod ?: static::SIGNATURE_METHOD_HMACSHA1;
    }

    /**
     * Getter: Retrieves option value for specified option key
     * @param string $option Option key to retrieve
     * @return mixed
     */
    public function getOption($option)
    {
        if (isset($this->_options[$option])) {
            return $this->_options[$option];
        }

        return null;
    }

    /**
     * Setter: Sets internal access token instance. The access token is stored as an array internally consisting of the token
     * and the associated secret.
     * <pre>
     *  accessToken = [
     *      accessToken => value,
     *      accessTokenSecret => value
     *  ]
     * </pre>
     *
     * @param string|array $token
     * @return $this
     */
    public function setAccessToken($token)
    {
        if (!$token) {
            $this->_accessToken = null;
            return $this;
        }

        if (is_string($token)) {
            $token = array('accessToken' => $token);
            if (is_array($this->_accessToken)) {
                $token = array_merge($this->_accessToken, $token);
            }
        }
        if (!is_array($token)) {
            throw new \InvalidArgumentException("Invalid access token provided. Expect string or array");
        }

        $this->_accessToken = $token;
        return $this;
    }

    /**
     * Setter: Sets the internal access token secret value
     * @param string $secret
     * @return $this
     */
    public function setAccessTokenSecret($secret)
    {
        if (!is_string($secret)) {
            throw new \InvalidArgumentException("Invalid access token secret provided. Expected string");
        }

        if (!is_array($this->_accessToken)) {
            $this->_accessToken = array();
        }
        $this->_accessToken['accessTokenSecret'] = $secret;
        return $this;
    }

    /**
     * Getter: Retrieves the internal access token detail specifier {@see setAccessToken}
     * @return array
     */
    public function getAccessToken()
    {
        return $this->_accessToken;
    }

    /**
     * Force authentication process
     *
     * @return bool
     */
    public function authenticate()
    {
        $this->_oa = null;
        $this->_oaData = array();
        $this->_accessToken = null;

        return $this->_establishAuthorization();
    }


    /**
     * Consumer Key
     * @var string
     */
    protected $_consumerKey;

    /**
     * Consumer Key Secret
     * @var string
     */
    protected $_consumerSecret;

    /**
     * Access Token
     * <pre>
     *  [
     *      accessToken => value,
     *      accessTokenSecret => value
     *  ]
     * </pre>
     *
     * @var array
     */
    protected $_accessToken;

    /**
     * Signature method to use
     * @var string
     */
    protected $_signatureMethod;

    /**
     * Username value to use for authentication
     * @var string
     */
    protected $_username;

    /**
     * Password value to use for authentication
     * @var string
     */
    protected $_password;

    /**
     * Last response from API interaction
     * @var array
     */
    protected $_lastResponse = null;

    /**
     * Error instance
     * <pre>
     *  [
     *      'code' => value,
     *      'description' => value
     *  ]
     * </pre>
     * @var bool|array
     */
    protected $_error = false;

    /**
     * Options array
     * @var array
     */
    protected $_options = array(
        'serviceRoot' => 'https://portal2.foreseeresults.com/services/',
        'requestTokenUri' => 'oauth/request_token',
        'accessTokenUri' => 'oauth/access_token',
        'authorizationUri' => 'oauth/user_authorization',
        'loginUri' => 'login',
        'sessionId' => null
    );

    /**
     * Internal OAuth instance
     * @var \OAuth\OAuth1\Service\Foresee
     */
    protected $_oa = null;

    /**
     * Internal OAuth cache data object
     * @var array
     */
    protected $_oaData = array();

    /**
     * Retrieves the OAuth instance, creates new instance and preforms authentication process should it be not done
     *
     * @return \OAuth\OAuth1\Service\Foresee
     * @codeCoverageIgnore
     */
    protected function _getOAuth()
    {
        if (!$this->_oa) {
            require_once __DIR__  . '/vendor/OAuth/bootstrap.php';

            $credentials = new \OAuth\Common\Consumer\Credentials(
                $this->getConsumerKey(),
                $this->getConsumerSecret(),
                'oob'
            );

            $cookieJar = tempnam(sys_get_temp_dir(), "acs_");
            $this->setOptions(array('cookieJar' => $cookieJar));

            $client = new \OAuth\Common\Http\Client\CurlClient('ACSOAuthLibrary');
            $client->setMaxRedirects(0);
            $client->setCurlParameters(array(
                CURLOPT_COOKIEFILE => $cookieJar,
                CURLOPT_COOKIEJAR => $cookieJar
            ));

            $storage = new \OAuth\Common\Storage\Memory();

            $signature = new \OAuth\OAuth1\Signature\Signature($credentials);
            $signature->setHashingAlgorithm($this->getSignatureMethod());

            $baseUri = new \OAuth\Common\Http\Uri\Uri($this->getOption('serviceRoot'));

            $this->_oa = new \OAuth\OAuth1\Service\Foresee(
                $credentials,
                $client,
                $storage,
                $signature,
                $baseUri
            );

            if ($this->_accessToken) {
                $token = $this->_oa->createAccessToken(array(
                    'oauth_token' => $this->_accessToken['accessToken'],
                    'oauth_token_secret' => $this->_accessToken['accessTokenSecret']
                ));
                $this->_oa->setAccessToken($token);
                $this->_oaData['accessToken'] = $token;
            }
        }

        return $this->_oa;
    }

    /**
     * Processes the authorization request and establishes credentials with the ACS API OAUTH System
     * @return bool
     */
    protected function _establishAuthorization()
    {
        $oa = $this->_getOAuth();

        if (isset($this->_oaData['accessToken'])) {
            return $this;
        }

        try {
            /**
             * *** REQUEST TOKEN
             */
            $this->_oaData['requestToken'] = $requestToken = $oa->requestRequestToken();

            if (!$requestToken) {
                // Add error return false
                $this->setError("INVALIDREQUESTTOKEN", "Unable to retrieve request token from service.");
                return false;
            }

            /**
             * @var \OAuth\Common\Http\Client\CurlClient $request
             */
            $request = $oa->getHttpClient();

            /**
             * LOGIN
             */
            $response = $request->retrieveResponse(
                $oa->getLoginEndpoint(),
                array(
                    'j_username' => $this->getUsername(),
                    'j_password' => $this->getPassword()
                ),
                array(),
                "POST"
            );

            $info = $request->getLastRequest('info');
            if ($info['http_code'] != 302) {
                $this->setError("COULDNOTLOGIN", "Unexpected response from the OAuth service, unable to authorize.");
                return false;
            }

            $redirectUrl = $info['redirect_url'];
            if (strpos($redirectUrl, '#loginfailed') !== FALSE) {
                $this->setError("INVALIDCREDENTIALS", "Unable to login, username and password are invalid.");
                return false;
            }

            /**
             * CHECK AUTHORIZATION / GET VERIFIER
             */
            $response = $request->retrieveResponse(
                $oa->getAuthorizationEndpoint(array(
                    'oauth_token' => $requestToken->getRequestToken()
                )),
                null,
                array(),
                "GET"
            );
            $last = $request->getLastRequest();
            $info = $last['info'];
            if ($info['http_code'] != '302') {
                $this->setError("COULDNOTGETACCESSTOKEN", "Unexpected response from the OAuth service, unable to authorize.");
                return false;
            }

            $url = $info['redirect_url'];
            if (strpos($url, '?') === FALSE) {
                $this->setError("COULDNOTGETACCESSTOKEN", "Unexpected response from the OAuth service, unable to authorize.");
                return false;
            }

            list($host, $url) = explode('?', $url, 2);
            parse_str($url, $values);
            if (!isset($values['oauth_verifier']) || strlen($values['oauth_verifier']) < 2) {
                $this->setError("COULDNOTFINDVERIFIER", "Unexpected response from OAuth service, unable to complete authorization.");
                return false;
            }

            $this->_oaData['oauth_verifier'] = $values['oauth_verifier'];

            /**
             * ACCESS TOKEN
             */
            $oa->setHttpClient($request);
            $accessToken = $oa->requestAccessToken(
                $requestToken->getRequestToken(),
                $this->_oaData['oauth_verifier'],
                $requestToken->getRequestTokenSecret()
            );
            if (!$accessToken) {
                $this->setError("COULDNOTGETACCESSTOKENNULL", "Unable to retrieve access token, please try again.");
                return false;

            }
            $this->setAccessToken($accessToken->getAccessToken())->setAccessTokenSecret($accessToken->getAccessTokenSecret());
            $this->_oaData['accessToken'] = $accessToken;
            return $this;
        } catch (Exception $e) {
            // Add error
            $this->setError("COULDNOTLOGIN", "Error while authorizing. " . $e->getMessage());
            return false;
        }
    }
}
